sway/input/cursor.c

```diff
@@ -827,22 +854,38 @@ static void handle_tool_button(struct wl_listener *listener, void *data) {
 				dispatch_cursor_button(cursor, &event->tablet->base,
 						event->time_msec, BTN_RIGHT, event->state);
 			}
-			cursor->tool_buttons++;
 			break;
 		case WLR_BUTTON_RELEASED:
-			if (cursor->tool_buttons == 1) {
+			if (cursor->tool_buttons <= 1) {
 				dispatch_cursor_button(cursor, &event->tablet->base,
 						event->time_msec, BTN_RIGHT, event->state);
 			}
-			cursor->tool_buttons--;
 			break;
 		}
 		wlr_seat_pointer_notify_frame(cursor->seat->wlr_seat);
-		return;
+	} else {
+		cursor->simulating_pointer_from_tool_button = false;
+
+		wlr_tablet_v2_tablet_tool_notify_button(sway_tool->tablet_v2_tool,
+			event->button, (enum zwp_tablet_pad_v2_button_state)event->state);
 	}
 
-	wlr_tablet_v2_tablet_tool_notify_button(sway_tool->tablet_v2_tool,
-		event->button, (enum zwp_tablet_pad_v2_button_state)event->state);
+	// Update tool button count.
+	switch (event->state) {
+	case WLR_BUTTON_PRESSED:
+		cursor->tool_buttons++;
+		break;
+	case WLR_BUTTON_RELEASED:
+		if (cursor->tool_buttons == 0) {
```

Comment:
To make this logic a little easier to read, can we write it as

```c
if (cursor->tool_buttons == 0) {
   ...
} else if (cursor->tool_buttons <= 1) {
   ...
} else {
   ...
}
```

I think writing it this way also makes the `<= 1` case impossible to ever be `<` (because we won't be decrementing when `cursor->tool_buttons = 0` to get into this branch in the first place)?

In which case we might just need

```c
if (cursor->tool_buttons == 0) {
    sway_log(SWAY_ERROR, "inconsistent tablet tool button events");
} else {
    cursor->tool_buttons--;
}
```

What do you think?

Reply:
Ahh, this feels obvious now that you point it out. Fixed.